### PR TITLE
GHC 9.6 support

### PR DESCRIPTION
**Pull request by christiaanb**
Add support for GHC 9.6, everything works ~~except the tests requiring hint which is broken on GHC 9.4~~.

### CODE REVIEW

File: clash-lib/src/Clash/Core/TyCon.hs [hunk 1]
```diff
@@ -119,12 +119,12 @@ mkKindTyCon name kind
...
-isTupleTyConLike nm = tupleName (nameOcc nm)
+isTupleTyConLike nm = tupleName (T.takeWhileEnd (/= '.') (nameOcc nm))
   where
-    tupleName nm'
-      | '(' <- T.head nm'
-      , ')' <- T.last nm'
-      = T.all (== ',') (T.init $ T.tail nm')
+    tupleName nm0
+      | Just ('(',nm1) <- T.uncons nm0
+      , Just (nm2,')') <- T.unsnoc nm1
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Core/TyCon.hs:126]:
```suggestion
      | Just ('(', nm1) <- T.uncons nm0
      , Just (nm2, ')') <- T.unsnoc nm1
```

Observe the style guide pls :eyes:

File: clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs [hunk 2]
```diff
@@ -4546,6 +4551,51 @@ ghcPrimStep tcm isSubj pInfo tys args mach = case primName pInfo of
...
+    -> reduce (valToTerm arg)
+  "GHC.TypeNats.withSomeSNat"
+    | Lit (NaturalLiteral n) : fun : _ <- args
+    , _:funTy:_ <- Either.rights (fst (splitFunForallTy ty))
+    , (tyView -> TyConApp snatTcNm _):_ <- Either.rights (fst (splitFunForallTy funTy))
+    , Just snatTc <- UniqMap.lookup snatTcNm tcm
+    , [snatDc] <- tyConDataCons snatTc
+    -> let nTy = LitTy (NumTy n)
+           snat = mkApps (Data snatDc) [Right nTy,Left (Literal (NaturalLiteral n))]
+           ret = mkApps (valToTerm fun) [Right nTy,Left snat]
```
Comment (martijnbastiaan) [inline clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs:4577]: This applies `withSomeSNat` to an `SNat`, but its type signature wants an `Maybe (SNat n)`. Is this right?

File: clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs [hunk 3]
```diff
@@ -4546,6 +4551,51 @@ ghcPrimStep tcm isSubj pInfo tys args mach = case primName pInfo of
...
+    | [arg] <- args
+    -> reduce (valToTerm arg)
+  "GHC.Float.$wproperFractionDouble"
+    | _:Lit (DoubleLiteral d):_ <- args
+    , [sty@(tyView -> TyConApp signedTcNm [nTy@(LitTy (NumTy kn))])] <- tys
+    , nameOcc signedTcNm == "Clash.Sized.Internal.Signed.Signed"
+    , (_,tyView -> TyConApp tupTcNm tyArgs) <- splitFunForallTy ty
+    , Just tupTc <- UniqMap.lookup tupTcNm tcm
+    , [tupDc] <- tyConDataCons tupTc
+    -> let (sn,d1) = reifyNat kn (\p -> first toInteger (op p (wordToDouble d)))
```
Comment (martijnbastiaan) [inline clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs:4589]: Could you please use spaces after `,` like our style guide suggests? It's happening a lot in this function.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -32,7 +32,7 @@ tests:
     strategy: depend
   parallel:
     matrix:
-      - GHC_VERSION: [8.8.4, 8.10.7, 9.2.5, 9.4.3]
+      - GHC_VERSION: [8.8.4, 8.10.7, 9.2.5, 9.4.3, 9.6.1]
```
Comment (martijnbastiaan) [inline .gitlab-ci.yml:35]: This change should also happen on GitHub Action. I think we also forgot for 9.4.3?

File: .ci/docker/build-and-publish-docker-image.sh [hunk 5]
```diff
@@ -14,8 +14,8 @@ elif [[ "$1" != "" ]]; then
   exit 1
 fi
 
-GHC_VERSIONS=(  "9.4.3"  "9.2.5"   "9.0.2"   "8.10.7"  "8.8.4"   "8.6.5")
-CABAL_VERSIONS=("3.8.1.0" "3.6.2.0" "3.4.0.0" "3.2.0.0" "3.2.0.0" "3.0.0.0")
+GHC_VERSIONS=(  "9.6.1"  "9.4.3"  "9.2.5"   "9.0.2"   "8.10.7"  "8.8.4"   "8.6.5")
+CABAL_VERSIONS=("3.10.1.0" "3.8.1.0" "3.6.2.0" "3.4.0.0" "3.2.0.0" "3.2.0.0" "3.0.0.0")
```
Comment (martijnbastiaan) [inline .ci/docker/build-and-publish-docker-image.sh:18]: This is getting quite a lot to test. I think we should only run tests on the last three versions for PRs, and run the full suite at night. (We should build everything on PRs though.) What do you think?

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]

**Review by martijnbastiaan — APPROVED** [review]
Just saw that my biggest point is moot, so this looks uncontroversial to me. Some tiny changes requested.

File: .gitlab-ci.yml (hunk 4 above)
Comment (christiaanb) [reply .gitlab-ci.yml:35]: There’s no stackage nightly yet for GHC 9.6, so no GHA for 9.6

File: .ci/docker/build-and-publish-docker-image.sh (hunk 5 above)
Comment (christiaanb) [reply .ci/docker/build-and-publish-docker-image.sh:18]: I agree, but perhaps make that change on a separate PR.

Comment (martijnbastiaan) [reply .ci/docker/build-and-publish-docker-image.sh:18]: Yeah, totally

**Pull request merged**